### PR TITLE
Computer action

### LINKED ISSUE #9: Management Module - Action Cards - "Skip"
* skip_card()

### LINKED ISSUE #11: Management Module - Action Cards - "Draw One"
* draw_one()

### LINKED ISSUE #12: Management Module - Action Cards - "Draw Two"
* draw_two()

### PR DESCRIPTION
**Pull request by HongBo713**
Add all function card into computer player, workable version of the release 2

### CODE REVIEW

File: include/computer_player.h [hunk 1]
```diff
@@ -141,6 +141,38 @@ Deck_t* pick_card(Card_t inputCard, Deck_t** hand_card);
...
+
+/**
+ * @brief Computer player discards a playable card, which include Skip and DRAW_ONE card.
+ *
+ */
+void computer_process_playable_card(Deck_t* playable_card, Deck_t** handcard);
+
+
+
+bool check_after_action(CardColor_e color, Deck_t** hand_card);
```
Comment (MiaYWei) [inline include/computer_player.h]: Redundant. Already have it in line 155. Let's remove it.

File: src/cards_management.c [hunk 2]
```diff
@@ -69,6 +69,7 @@ int initialize_cards(void)
     card.name = WILD_DRAW_TWO;
     result += add_card_at_beginning(&g_draw_pile, card);
     result += add_card_at_beginning(&g_draw_pile, card);
+ /*   shuffle_cards();*/
```
Comment (MiaYWei) [inline src/cards_management.c]: Still need it? If yes, then uncomment it. Otherwise, please delete it.

File: src/computer_player.c [hunk 3]
```diff
@@ -20,11 +20,12 @@
...
+#include "../include/game.h"
 
 #define hand_MAX 20
 #define plarCardNumber 10
 
-///* (not used for now) */
+
 //struct CARD record_array[MAX]; // array for recording all the cards
 //int record_size = 0, handcard_size = 0;
 //
```
Comment (MiaYWei) [inline src/computer_player.c]: Let's delete the unused code.

File: src/computer_player.c [hunk 4]
```diff
@@ -320,16 +321,27 @@ casenumber_e pick_case(Card_t card, Deck_t** hand_card)
...
+    //if (card.name == DRAW_ONE) {
+    //    return CASE5;
+    //}
+    //// draw two card if it is WILD DRAW TWO 
+    //else if (card.name == WILD_DRAW_TWO) {
+    //    return CASE6;
+    //}
+    //else if (card.name == SKIP) {
+    //    return CASE7;
+    //}
```
Comment (MiaYWei) [inline src/computer_player.c]: Let's delete the unused code.

File: src/computer_player.c [hunk 5]
```diff
@@ -400,40 +414,40 @@ Deck_t* pick_card(Card_t inputCard, Deck_t** hand_card)
     Deck_t* temp = *hand_card;
     enum casenumber caseNum = pick_case(inputCard, hand_card);
     most_color = find_most_color(hand_card);
-    //if (temp->next->next = NULL) {
-    //    playableCard.name = WILD;
-    //    playCard = find_address(hand_card, playableCard);
-    //    playCard->card.color = most_color;
-    //}
+
```
Comment (MiaYWei) [inline src/computer_player.c:388]: Let's delete the unused code.

File: include/computer_player.h [hunk 6]
```diff
@@ -9,7 +9,7 @@
 #define hand_MAX 20
 #define plarCardNumber 10
 
-typedef enum casenumber { CASE1, CASE2, CASE3, CASE4 } casenumber_e;
+typedef enum casenumber { CASE1, CASE2, CASE3, CASE4, CASE5, CASE6, CASE7 } casenumber_e;
```
Comment (MiaYWei) [inline include/computer_player.h]: Seems like we don't use CASE5, CASE6, and CASE7. Let's remove them.

**Review by MiaYWei — APPROVED** [review]
Great work! Let's cleanup the code based on my suggested code comments.

File: include/computer_player.h [hunk 7]
```diff
@@ -141,6 +141,38 @@ Deck_t* pick_card(Card_t inputCard, Deck_t** hand_card);
...
+ *
+ * @param        0 - draw one card and evalue if it's a valid card to play
+ *               1 - draw one card and end turn directly
+ *               2 - draw two card and end turn directly
+ *
+ * @return int   0 - Discarding card is successful, end of turn, game continues.
+ *               1 - No playable card to discard, end of turn, game continues.
+ *               2 - Invalid player.
+ */
+bool check_after_action(CardColor_e color, Deck_t** hand_card);
```
Comment (krithigasr) [inline include/computer_player.h:151]: Can you change the function name to match its description? Can't really understand just by looking at the function name.

File: src/computer_player.c [hunk 8]
```diff
@@ -567,12 +590,71 @@ Deck_t* play_card(const Deck_t* cardAddress, Deck_t** head)
         if (temp == NULL)
             return NULL;
         prev->next = temp->next;
-        printf("Successfully deleted  card (%s, %s). \n", CARD_COLOR_STRING[temp->card.color], CARD_NAME_STRING[temp->card.name]);
+        //printf("Successfully deleted  card (%s, %s). \n", CARD_COLOR_STRING[temp->card.color], CARD_NAME_STRING[temp->card.name]);
```
Comment (krithigasr) [inline src/computer_player.c:570]: I just noticed the function description has mismatched parameter values. Please change

File: src/computer_player.c [hunk 9]
```diff
@@ -567,12 +590,71 @@ Deck_t* play_card(const Deck_t* cardAddress, Deck_t** head)
...
         return temp;
 
     }
 }
 
+/**
+ * @brief No playable card, computer player reuqests a new card
```
Comment (krithigasr) [inline src/computer_player.c]: Can you change this to "Computer requests a new card when it has no playable card on hand"? Lets be more descriptive.

File: src/computer_player.c [hunk 10]
```diff
@@ -567,12 +590,71 @@ Deck_t* play_card(const Deck_t* cardAddress, Deck_t** head)
...
+
+    end_turn(COMPUTER);
+    return result;
+}
+
+
+/**
+ * @brief Computer player discards a playable card, which include Skip and DRAW_ONE card.
+ *
+ */
```
Comment (krithigasr) [inline src/computer_player.c:610]: Can you add in/out param descriptions here?

File: src/game.c [hunk 11]
```diff
@@ -38,8 +38,8 @@ void start_new_game(void)
     printf("Human Player Deck: ");
     display_cards_list((const Deck_t*)g_players[HUMAN].cards_on_hand);
     //For Testing only
-    //printf("Computer Player Deck: ");
-    //display_cards_list((const Deck_t*)g_players[COMPUTER].cards_on_hand);
+    printf("Computer Player Deck: ");
```
Comment (krithigasr) [inline src/game.c]: Please comment this.

File: src/game.c [hunk 12]
```diff
@@ -38,8 +38,8 @@ void start_new_game(void)
...
+    display_cards_list((const Deck_t*)g_players[COMPUTER].cards_on_hand);
```
Comment (krithigasr) [inline src/game.c]: This too

File: src/game.c [hunk 13]
```diff
@@ -191,14 +191,14 @@ bool if_end_game(PlayerType_e player)
...
     int ret = computer_take_turn();
-    if ((0 == ret) || (1 == ret))
-    {
-        end_turn(COMPUTER);
-    }
-    else
-    {
-        printf("Error: Not Computer's turn now.\n");
-    }
+    //if ((0 == ret) || (1 == ret))
```
Comment (krithigasr) [inline src/game.c]: Can you remove this if unused?

**Review by krithigasr — COMMENTED** [review]
General comment:
Let's use function names which are self explanatory so that everyone can understand without having to go through the code.

**Review by aditibhardwaj98 — APPROVED** [review]

**Pull request merged**